src/borg/archive.py

```diff
@@ -496,9 +499,21 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
                     return
                 # Extract chunks, since the item which had the chunks was not extracted
             with backup_io():
-                fd = open(path, 'wb')
+                fd = open(path, 'ab')
             with fd:
-                ids = [c.id for c in item.chunks]
+                chunks = item.chunks
+                if complete_partial:
+                    with backup_io():
+                        fd.seek(0, io.SEEK_END)
+                        existing_length = fd.tell()
+                        # Slice chunks by current length of the existing file.
+                        chunks, prefix_length = slice_chunks(chunks, maximum_length=existing_length)
+                        # We don't bother extracting fractional chunks. Just seek to a chunk boundary.
+                        fd.seek(prefix_length)
+                        fd.truncate()
+                    if pi:
+                        pi.show(increase=prefix_length)
+                ids = [c.id for c in chunks]
                 for _, data in self.pipeline.fetch_many(ids, is_preloaded=True):
```

Comment:
not sure if this causes an issue with the preloaded chunks:
iirc it preloads all chunks of a file when processing its item, but then it fetches either none (existing, full file) or only some (existing, partial file), before it goes into fetch-all mode (non-existing files).